File: android/app/src/main/kotlin/com/rpm/app/ui/feature/patients/SupportedVitals.kt
package com.rpm.app.ui.feature.patients

import com.rpm.app.data.remote.dto.VitalRecordDto
import com.rpm.app.data.remote.dto.VitalRecordLatestDto
import com.rpm.app.data.signalr.RealTimeVitals

/**
 * Vitals the Galaxy Watch can share (Samsung Health Sensor SDK + platform sensors).
 * @see https://developer.samsung.com/health/sensor/guide/data-specifications.html
 */
object SupportedVitals {

    data class MetricDef(
        val label: String,
        val unit: String = "",
        val hint: String? = null,
        val getValue: (RealTimeVitals?) -> String,
    )

    val liveMetrics: List<MetricDef> = listOf(
        MetricDef("Heart Rate", "bpm") { v -> v?.heartRateBpm?.toInt()?.toString() ?: "--" },
        MetricDef("HRV", "ms") { v -> v?.hrvMs?.toInt()?.toString() ?: "--" },
        MetricDef("SpO₂", "%", hint = "On-demand") { v -> v?.spO2Percent?.let { "%.1f".format(it) } ?: "--" },
        MetricDef("Skin Temp.", "°C") { v -> v?.skinTemperatureC?.let { "%.1f".format(it) } ?: "--" },
        MetricDef("Ambient Temp.", "°C") { v -> v?.ambientTemperatureC?.let { "%.1f".format(it) } ?: "--" },
        MetricDef("Stress", "/100") { v -> v?.stressScore?.toInt()?.toString() ?: "--" },
        MetricDef("Steps", "today") { v -> v?.stepsCount?.toString() ?: "--" },
        MetricDef("Calories", "kcal") { v -> v?.caloriesBurned?.let { "%.0f".format(it) } ?: "--" },
        MetricDef("Fall Detection", "") { v ->
            when (v?.fallDetected) { true -> "Alert!"; false -> "Safe"; null -> "--" }
        },
        MetricDef("Watch Status", "") { v ->
            when {
                v == null -> "--"
                isWearing(v) -> "On Wrist"
                else -> "Off Wrist"
            }
        },
        MetricDef("Body Fat", "%", hint = "On-demand") { v -> v?.bodyFatPercent?.let { "%.1f".format(it) } ?: "--" },
        MetricDef("ECG Avg HR", "bpm", hint = "On-demand") { v -> v?.ecgAvgHeartRateBpm?.toInt()?.toString() ?: "--" },
    )

    fun isWearing(v: RealTimeVitals): Boolean =
        v.isWearing || v.heartRateBpm != null || v.spO2Percent != null
            || v.skinTemperatureC != null || v.stressScore != null || v.hrvMs != null

    fun isWearing(record: VitalRecordDto): Boolean =
        record.isWearing || record.heartRateBpm != null || record.spO2Percent != null
            || record.skinTemperatureC != null || record.stressScore != null || record.hrvMs != null

    fun buildSummary(v: VitalRecordLatestDto): String = buildString {
        v.heartRateBpm?.let { append("HR: ${it.toInt()} bpm") }
        v.spO2Percent?.let {
            if (isNotEmpty()) append("  •  ")
            append("SpO₂: ${it.toInt()}%")
        }
        v.skinTemperatureC?.let {
            if (isNotEmpty()) append("  •  ")
            append("Skin: %.1f °C".format(it))
        } ?: v.temperatureC?.let {
            if (isNotEmpty()) append("  •  ")
            append("%.1f °C".format(it))
        }
        v.stressScore?.let {
            if (isNotEmpty()) append("  •  ")
            append("Stress: ${it.toInt()}")
        }
        if (isEmpty()) append("No vitals recorded yet")
    }

    /** Rows for history / detail — only non-null supported fields. */
    fun historyRows(record: VitalRecordDto): List<Pair<String, String>> = buildList {
        record.heartRateBpm?.let { add("Heart Rate" to "${it.toInt()} bpm") }
        record.spO2Percent?.let { add("SpO₂" to "${it.toInt()}%") }
        record.skinTemperatureC?.let { add("Skin Temp." to "%.1f °C".format(it)) }
        record.ambientTemperatureC?.let { add("Ambient Temp." to "%.1f °C".format(it)) }
        record.hrvMs?.let { add("HRV" to "${it.toInt()} ms") }
        record.stressScore?.let { add("Stress" to "${it.toInt()} / 100") }
        record.stepsCount?.let { add("Steps" to "$it") }
        record.caloriesBurned?.let { add("Calories" to "%.0f kcal".format(it)) }
        record.bodyFatPercent?.let { add("Body Fat" to "%.1f%%".format(it)) }
        record.ecgAvgHeartRateBpm?.let { add("ECG Avg HR" to "${it.toInt()} bpm") }
        add("Watch" to if (isWearing(record)) "On wrist" else "Off wrist")
    }
}

File: web-app/src/utils/supportedVitals.js
/**
 * Vitals the Galaxy Watch app can share (Samsung Health Sensor SDK + platform sensors).
 * @see https://developer.samsung.com/health/sensor/guide/data-specifications.html
 */

export const SDK_SENSOR_INFO = {
  continuous: [
    { name: 'Heart rate + HRV', tracker: 'HEART_RATE_CONTINUOUS', note: 'Galaxy Watch4+' },
    { name: 'Skin & ambient temperature', tracker: 'SKIN_TEMPERATURE_CONTINUOUS', note: 'Galaxy Watch5+' },
    { name: 'Stress (EDA)', tracker: 'EDA_CONTINUOUS', note: 'Galaxy Watch8+' },
  ],
  onDemand: [
    { name: 'SpO₂', tracker: 'SPO2_ON_DEMAND', note: 'Auto every ~3 min' },
    { name: 'Body fat (BIA)', tracker: 'BIA_ON_DEMAND', note: 'Manual on watch' },
    { name: 'ECG avg heart rate', tracker: 'ECG_ON_DEMAND', note: 'Manual on watch' },
  ],
  platform: [
    { name: 'Steps & calories', note: 'Wear OS step counter' },
    { name: 'Fall detection', note: 'Accelerometer' },
    { name: 'Watch on-wrist', note: 'Samsung off-body + live vitals' },
  ],
}

/** Metrics shown in the live grid — only what the watch can provide. */
export const SUPPORTED_METRIC_DEFS = [
  { key: 'heartRateBpm', label: 'Heart rate', unit: 'bpm', tone: 'accent', digits: 0, source: 'sdk' },
  { key: 'hrvMs', label: 'HRV', unit: 'ms', tone: 'violet', digits: 0, source: 'sdk' },
  { key: 'spO2Percent', label: 'SpO₂', unit: '%', tone: 'teal', digits: 1, source: 'sdk', onDemand: true },
  { key: 'skinTemperatureC', label: 'Skin temp.', unit: '°C', tone: 'blue', digits: 1, source: 'sdk' },
  { key: 'ambientTemperatureC', label: 'Ambient temp.', unit: '°C', tone: 'ink', digits: 1, source: 'sdk' },
  { key: 'stressScore', label: 'Stress', unit: '/100', tone: 'violet', digits: 0, source: 'sdk' },
  { key: 'stepsCount', label: 'Steps', unit: 'today', tone: 'ink', digits: 0, source: 'platform' },
  { key: 'caloriesBurned', label: 'Calories', unit: 'kcal', tone: 'ink', digits: 0, source: 'platform' },
  { key: 'fallDetected', label: 'Fall detection', unit: '', tone: 'danger', boolean: true, source: 'platform' },
  { key: 'isWearing', label: 'Watch status', unit: '', tone: 'teal', wearing: true, source: 'platform' },
  { key: 'bodyFatPercent', label: 'Body fat', unit: '%', tone: 'ink', digits: 1, source: 'sdk', onDemand: true },
  { key: 'ecgAvgHeartRateBpm', label: 'ECG avg HR', unit: 'bpm', tone: 'ink', digits: 0, source: 'sdk', onDemand: true },
]

export const HISTORY_COLUMNS = [
  { key: 'heartRateBpm', label: 'HR', digits: 0 },
  { key: 'spO2Percent', label: 'SpO₂', digits: 1 },
  { key: 'skinTemperatureC', label: 'Skin °C', digits: 1 },
  { key: 'ambientTemperatureC', label: 'Amb. °C', digits: 1 },
  { key: 'hrvMs', label: 'HRV', digits: 0 },
  { key: 'stressScore', label: 'Stress', digits: 0 },
  { key: 'stepsCount', label: 'Steps', digits: 0 },
]
